### DISCUSSION TITLE
Real-time robot state information transfer to external program

### DISCUSSION THREAD

**Discussion by Noeulleaf** [UniversalRobots/Universal_Robots_ROS2_Driver#943] [Q&A]
Hello,

I am currently working on a program that uses the tool_communication_node to transmit the robot's state information in real time to an external program and display it in a Windows-based program.

After reviewing the tool_communication_node.py file, it seems that the node does not directly handle the message format. I would like to confirm if the node transmits messages in the rtde_output_recipe.txt format.

thank you.

**Comment by RobertWilbrandt** [comment]
That is not what the tool communication script does. It is used to forward the RS-485 interface at the end-effector flange to the PC, allowing you to e.g. directly control grippers. More information on it can be found [in the docs](https://docs.ros.org/en/ros2_packages/rolling/api/ur_robot_driver/setup_tool_communication.html).

The `rtde_output_recipe.txt` file is used by our driver to communicate with the arm using [RTDE](https://www.universal-robots.com/articles/ur/interface-communication/real-time-data-exchange-rtde-guide/). You should be able to just use our ROS interfaces (`/joint_states`, `/force_torque_sensor_broadcaster/wrench`, ...) to get all the data you need without directly accessing the RTDE layer.
